src/bits/bit_array.rs

```diff
@@ -0,0 +1,181 @@
+use super::{BitArray, BooleanOps};
+use bitvec::prelude::{BitArr, Lsb0};
+use std::{
+    fmt::Debug,
+    ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXor, BitXorAssign, Index, Not},
+};
+
+/// Bit store size of `XorSecretShare` in bytes.
+const BIT_STORE_SIZE_IN_BYTES: usize = 4;
+type U8_4 = BitArr!(for BIT_STORE_SIZE_IN_BYTES * 8, in u8, Lsb0);
+
+/// 32-bit array of bits. Similar to `u32`, but supports boolean algebra, and
+/// provides access to individual bits via index.
+///
+/// Bits are stored in the Little-Endian format. Accessing the first element
+/// like `s[0]` will return the LSB.
+///
+/// ## Example
+/// ```
+/// use raw_ipa::bits::BitArray32;
+///
+/// let s = BitArray32::from(2_u128);
+/// let b0 = s[0];
+/// let b1 = s[1];
+///
+/// assert_eq!(b0, 0);
+/// assert_eq!(b1, 1);
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BitArray32(U8_4);
+
+impl BitArray for BitArray32 {
+    const SIZE_IN_BYTES: usize = BIT_STORE_SIZE_IN_BYTES;
+
+    #[allow(dead_code)]
+    const ZERO: Self = Self(U8_4::ZERO);
+}
+
+impl BooleanOps for BitArray32 {}
+
+impl BitAnd for BitArray32 {
+    type Output = Self;
+    fn bitand(self, rhs: Self) -> Self::Output {
+        Self(self.0 & rhs.0)
+    }
+}
+
+impl BitAndAssign for BitArray32 {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self.0.as_mut_bitslice() &= rhs.0;
+    }
+}
+
+impl BitOr for BitArray32 {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self(self.0 | rhs.0)
+    }
+}
+
+impl BitOrAssign for BitArray32 {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self.0.as_mut_bitslice() |= rhs.0;
+    }
+}
+
+impl BitXor for BitArray32 {
+    type Output = Self;
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        Self(self.0 ^ rhs.0)
+    }
+}
+
+impl BitXorAssign for BitArray32 {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        *self.0.as_mut_bitslice() ^= rhs.0;
+    }
+}
+
+impl Not for BitArray32 {
+    type Output = Self;
+    fn not(self) -> Self::Output {
+        Self(!self.0)
+    }
+}
+
+impl<T: Into<u128>> From<T> for BitArray32 {
```

Comment:
I think this is fallible conversion, is there a reason for having it? we could implement `From<u32>` that is guaranteed to be infallible. If we really need to, `TryFrom<u128>` can be implemented

Reply:
I think that I agree.  Though that might mean changing how we do fields as well.  That is, your common definition has `TryFrom<u128>` rather than `From<u128>`.  That should still work, though we'll be using `unwrap()` more often.

Reply:
I'd prefer `TryFrom<u128>` if we were creating a public library, but for the internal use I chose fewer typings. The code below could be simplified a lot by having this conversion.

```
    let xor_share = XorReplicated::new(
        u64::try_from(b_bits_left & u128::from(u64::MAX)).unwrap(),
        u64::try_from(b_bits_right & u128::from(u64::MAX)).unwrap(),
    );
```